### ISSUE TITLE
spawn phpcs ENOENT

### ISSUE THREAD

**Issue by hylle** [AtomLinter/linter-phpcs#32]
After the 1.0.1 update I get the following error message.

```
Error: spawn phpcs ENOENT
  at exports._errnoException (util.js:734:11)
  at Process.ChildProcess._handle.onexit (child_process.js:1035:32)
  at child_process.js:1127:20
  at process._tickCallback (node.js:357:13)
```

This is running the latest OS X 10.10.4.
Before the following phpcs path worked for me: "/usr/local/bin", also tried the following after the update "/usr/local/bin/phpcs", both return the same error.

**Comment by shaunburdick** [comment]
I get the same error.

**Comment by kingkool68** [comment]
Ditto. Just upgraded to 1.0.1 and now this breaks. Have the exact path to phpcs in the settings which i got from running `which phpcs` and it is still broken. 

Running OS X 10.10.4 if that matters.

**Comment by sdustinh** [comment]
Same on Windows 7

**Comment by luishdez** [comment]
Same here on OS X yosemite

**Comment by madhancock** [comment]
The errors are triggered by the config file changes:

- standard => codeStandardOrConfigFile
- phpcsExecutablePath => executablePath

You can either move/copy the values using the setting UI or by editing the config file itself.

My config file now contains both new and old names:

```
"linter-phpcs":
    phpcsExecutablePath: "/usr/local/opt/php56/bin/phpcs"
    standard: "PSR2"
    codeStandardOrConfigFile: "PSR2"
    executablePath: "/usr/local/opt/php56/bin/phpcs"
```

No more issues.

**Comment by steelbrain** [comment]
@madhancock the `phpcsExecutablePath` was actually replaced by `executablePath` and `standard` by `codeStandardOrConfigFile` in the latest update

**Comment by steelbrain** [comment]
and people, atom has a bug where the configs won't be updated after a package upgrade so you might wanna try
```
mv ~/.atom ~/.atom-bkp
apm install linter
apm install linter-phpcs
atom .
```
and see if it's a bug in this package or a cache issue

**Comment by madhancock** [comment]
@steelbrain Thats what I said, and why I included brief instructions to assist those with the problem in solving it.

Side question, is it not possible to support both config directives?

**Comment by steelbrain** [comment]
@madhancock It is. but not the neatest way to solve this problem.

**Comment by kingkool68** [comment]
Props to @madhancock :+1: that worked!

Software sucks!

**Comment by guillempascual** [comment]
In Win 8.1 the problem is command "spawn" does not exist.
Clearly it's not been tested, it worked well for me until the latest update.

spawn C:\php-5.5.3-Win32-VC11-x64\phpcs ENOENT
Error: spawn C:\php-5.5.3-Win32-VC11-x64\phpcs ENOENT
    at exports._errnoException (util.js:734:11)
    at Process.ChildProcess._handle.onexit (child_process.js:1035:32)
    at child_process.js:1127:20
    at process._tickCallback (node.js:357:13)

**Comment by steelbrain** [comment]
@guillempascual `spawn` isn't really a command that's executed, it's a nodejs builtin function, just like the trace indicates, also `ENOENT` is a node js internal error, it means that either the file was not found or was not executable.
In your case, You need to append `.exe` in front of it's name on windows.

**Comment by guillempascual** [comment]
@steelbrain thank you, but what do you mean?

I've got phpcs on my path , and the command "phpcs" works fine from the command line in any location. No need for phpcs.exe. 

Or did you mean anything else?

**Comment by steelbrain** [comment]
It works fine from the command line I agree, but there are two ways to spawn processes in node, You either spawn a command line, and then execute that command in it, or spawn the process directly, spawning the process directly needs EXACT same path, it doesn't append `.exe` for you automatically like the cmd one. and We're using it for it's flexibility

**Comment by guillempascual** [comment]
@steelbrain thank you.

I've got phpcs installed through PEAR.

This is phpcs.bat


This is the exact same path, C:\php-5.5.3-Win32-VC11-x64\phpcs\pear\phpcs.bat, but still refuses to run.



spawn C:\php-5.5.3-Win32-VC11-x64\phpcs\pear\phpcs.bat ENOENT

**Comment by steelbrain** [comment]
That is super-weird. Can you please downgrade through the version and tell us in which one it works for you?
Also can you execute this in your dev tools console and show us the output
```js
var path = 'C:\php-5.5.3-Win32-VC11-x64\phpcs\pear\phpcs.bat'
var cp = require('child_process')
cp.exec(path, function(err, data){
  console.log(err, data)
})
```

**Comment by guillempascual** [comment]
@steelbrain 

As to snippet execution the console:
var path = 'C:\php-5.5.3-Win32-VC11-x64\phpcs\pear\phpcs.bat'
var cp = require('child_process')
cp.exec(path, function(err, data){
  console.log(err, data)
})
ReferenceError: require is not defined

As to the downgrade: I have to find out how to downgrade, I'll tell you once I'm done.

Thank you for your help!!

Guillem

**Comment by steelbrain** [comment]
Are you actually executing it in your atom's dev tools and not chrome's dev tools?

**Comment by guillempascual** [comment]
I'm actually executing in Firefox Developer Edition 41.0a2 (2015-07-12).

Shall I execute in any other place?

**Comment by steelbrain** [comment]
You should open dev tools of atom itself and execute there.

**Comment by guillempascual** [comment]
Done.
It seems like the program is peeling of the '\' so the route can't be found.


var path = 'C:\php-5.5.3-Win32-VC11-x64\phpcs\pear\phpcs.bat'
var cp = require('child_process')
cp.exec(path, function(err, data){
console.log(err, data)
})

ChildProcess {domain: null, _events: Object, _eventsCount: 2, _maxListeners: undefined, _closesNeeded: 3…}
VM1082:5 Error: Command failed: C:\WINDOWS\system32\cmd.exe /s /c "C:php-5.5.3-Win32-VC11-x64phpcspearphpcs.bat"
"C:php-5.5.3-Win32-VC11-x64phpcspearphpcs.bat" no se reconoce como un comando interno o externo,
programa o archivo por lotes ejecutable.
 {killed: false, code: 1, signal: null, cmd: "C:\WINDOWS\system32\cmd.exe /s /c "C:php-5.5.3-Win32-VC11-x64phpcspearphpcs.bat"", stack: (...)…} ""

**Comment by steelbrain** [comment]
Ah right, You'll actually have to escape `\` with `\\` in javascript.

**Comment by guillempascual** [comment]
Yeah I already did

VM1128:5 Error: Command failed: C:\WINDOWS\system32\cmd.exe /s /c "C:\php-5.5.3-Win32-VC11-x64\phpcs\pearphpcs.bat"
El sistema no puede encontrar la ruta especificada.
 {killed: false, code: 1, signal: null, cmd: "C:\WINDOWS\system32\cmd.exe /s /c "C:\php-5.5.3-Win32-VC11-x64\phpcs\pearphpcs.bat"", stack: (...)…} ""

**Comment by steelbrain** [comment]
Italian? :sweat_smile: what does it mean?

**Comment by guillempascual** [comment]
spanish it means "the system can't find the route"

**Comment by steelbrain** [comment]
Have you confirmed that the path is correct?

**Comment by guillempascual** [comment]
Yes absolutely
C:\php-5.5.3-Win32-VC11-x64\pear> phpcs.bat --version
PHP_CodeSniffer version 2.3.3 (stable) by Squiz (http://www.squiz.net)

**Comment by guillempascual** [comment]
var path = 'C:\\php-5.5.3-Win32-VC11-x64\\phpcs\\pear\\phpcs.bat'
var cp = require('child_process')
cp.exec(path, function(err, data){
console.log(err, data)
})
ChildProcess {domain: null, _events: Object, _eventsCount: 2, _maxListeners: undefined, _closesNeeded: 3…}
VM1228:5 Error: Command failed: C:\WINDOWS\system32\cmd.exe /s /c "C:\php-5.5.3-Win32-VC11-x64\phpcs\pear\phpcs.bat"
El sistema no puede encontrar la ruta especificada.
 {killed: false, code: 1, signal: null, cmd: "C:\WINDOWS\system32\cmd.exe /s /c "C:\php-5.5.3-Win32-VC11-x64\phpcs\pear\phpcs.bat"", stack: (...)…} ""

**Comment by guillempascual** [comment]
System can't find specified route

**Comment by guillempascual** [comment]
Until this morning everything was working fine. I've been with linter-phpcs since ever.

**Comment by guillempascual** [comment]
Thank you for helping.

**Comment by guillempascual** [comment]
OK the problem in the snippet was in the route you gave me

var path = 'C:\\php-5.5.3-Win32-VC11-x64\\pear\\phpcs.bat'
var cp = require('child_process')
cp.exec(path, function(err, data){
console.log(err, data)
})
ChildProcess {domain: null, _events: Object, _eventsCount: 2, _maxListeners: undefined, _closesNeeded: 3…}
VM1233:5 null "

**Comment by steelbrain** [comment]
Now that's some process, now can you try this
```
var path = 'C:\php-5.5.3-Win32-VC11-x64\pear\phpcs.bat'
var cp = require('child_process')
cp.spawn(path)
```
and see if you get any error

**Comment by guillempascual** [comment]
var path = 'C:\php-5.5.3-Win32-VC11-x64\pear\phpcs.bat'
var cp = require('child_process')
cp.spawn(path)
ChildProcess {domain: null, _events: Object, _eventsCount: 0, _maxListeners: undefined, _closesNeeded: 3…}
events.js:141 Uncaught Error: spawn C:php-5.5.3-Win32-VC11-x64pearphpcs.bat ENOENT

**Comment by guillempascual** [comment]
you see the route pass to patht is not OK

**Comment by steelbrain** [comment]
Ah right, forgot to tell you to escape it again.

**Comment by guillempascual** [comment]
Sorry my mistake also.

Now
var path = 'C:\\php-5.5.3-Win32-VC11-x64\\pear\\phpcs.bat'
var cp = require('child_process')
cp.spawn(path)
child_process.js:1145 Uncaught Error: spawn UNKNOWNcode: "UNKNOWN"errno: "UNKNOWN"message: "spawn UNKNOWN"stack: (...)get stack: function () { [native code] }set stack: function () { [native code] }syscall: "spawn"__proto__: d

**Comment by guillempascual** [comment]
More detailed

var path = 'C:\\php-5.5.3-Win32-VC11-x64\\pear\\phpcs.bat'
var cp = require('child_process')
cp.spawn(path)
child_process.js:1145 Uncaught Error: spawn UNKNOWNcode: "UNKNOWN"errno: "UNKNOWN"message: "spawn UNKNOWN"stack: (...)get stack: function () { [native code] }set stack: function () { [native code] }syscall: "spawn"__proto__: dutil.js:734 exports._errnoExceptionchild_process.js:1145 ChildProcess.spawnchild_process.js:977 exports.spawnVM1313:4 (anonymous function)VM1033:847 InjectedScript._evaluateOnVM1033:780 InjectedScript._evaluateAn

**Comment by guillempascual** [comment]
In the console there is actually the double \\

**Comment by guillempascual** [comment]
gotta go we keep in contact thank you

**Comment by steelbrain** [comment]
I am actually busy working on a project, You can try pinging people in #linter of Atom's slack.
The url to join is https://atom-slack.herokuapp.com

**Comment by guillempascual** [comment]
OK. So I guess the trace of the snippet didn't reveal anything. I'll try the Atom's slack. Thank you.
